File: 剑指offer/JZ57二叉树的下一个结点.py
# -*- coding:utf-8 -*-
# class TreeLinkNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#         self.next = None
class Solution:
    def GetNext(self, pNode):
        # write code here
        if not pNode: return None
        if pNode.right:
            pNode = pNode.right
            while pNode.left:
                pNode = pNode.left
            return pNode
        while pNode.next:
            node = pNode.next
            if node.left == pNode:
                return node
            pNode = pNode.next
        return None

File: 剑指offer/JZ57二叉树的下一个结点.js
/*function TreeLinkNode(x){
    this.val = x;
    this.left = null;
    this.right = null;
    this.next = null;
}*/
function GetNext(pNode) {
    // write code here
    if (!pNode) return pNode;
    if (pNode.right) {
        pNode = pNode.right;
        while (pNode.left) {
            pNode = pNode.left;
        }
        return pNode;
    }

    while (pNode.next) {
        let node = pNode.next;
        if (node.left === pNode) {
            return node;
        }
        pNode = pNode.next;
    }
    return null;
}
module.exports = {
    GetNext: GetNext
};